distribution/xfer/download.go

```diff
@@ -414,11 +418,15 @@ func (ldm *LayerDownloadManager) makeDownloadFuncFromDownload(descriptor Downloa
 			}
 			defer layerReader.Close()
 
-			var src *distribution.Descriptor
-			if fs, ok := l.(layer.ForeignSourcer); ok {
```

Comment:
Yes, I think that is a problem. It should use the descriptor from "descriptor". You are correct that these may collide if the layers have the same parent chain. That's an unavoidable consequence of having a many-to-one mapping between URLs/descriptors and layers on disk.